### PR TITLE
Feature/ops 1722 build and deploy antivirus

### PR DESCRIPTION
**Pull request by RalfNellesCap**
# Description

Add CI Pipeline with Github Action, Deploy with sc-app-ci
## Links to Tickets or other pull requests

https://github.com/hpi-schul-cloud/schulcloud-server/pulls?q=is%3Apr+is%3Aopen+OPS-1559
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1874
## Changes

Enables CI to project with build and push to Dockerhub.
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1874
## Datasecurity <sub><sup>details [on Confluence](https://docs.schul-cloud.org/x/2S3GBg)</sup></sub>

N/A
## Deployment

Deployment will be added in another task with sc-app-ci
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1559
## New Repos, NPM pakages or vendor scripts

N/A
## Screenshots of UI changes

N/A
## Approval for review

- All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### Link to Definiton of Done

More and detailed information on the _definition of done_ can be found [on Confluence](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762)

### CODE REVIEW

File: .github/workflows/build_push_service.yml [hunk 1]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+        uses: actions-ecosystem/action-regex-match@v2
+        id: branch_prefix
+        with:
+          text: ${{ env.BRANCH_NAME }}
+          regex: '^([a-zA-Z]+)\/[A-Z]+-[0-9]+-[a-zA-Z_]+$'
+          flags: gm
+      - name: Set env var JIRA BRANCH_PREFIX
+        run: |
+          echo "BRANCH_PREFIX is ${{ steps.branch_prefix.outputs.group1 }}"
+          echo BRANCH_PREFIX=${{ steps.branch_prefix.outputs.group1 }} >> $GITHUB_ENV
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:81]: Why is it put to ENV vars if is at the output?

File: .github/workflows/build_push_service.yml [hunk 2]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+        uses: actions-ecosystem/action-regex-match@v2
+        id: jira
+        with:
+          text: ${{ env.BRANCH_NAME }}
+          regex: '^hotfix|feature\/([A-Z]+-[0-9]+)-[a-zA-Z_]+$'
+          flags: gm
+      - name: Set env var JIRA_TicketID
+        run: |
+          echo "JIRA_TicketID is ${{ steps.jira.outputs.group1 }}"
+          echo JIRA_TicketID=${{ steps.jira.outputs.group1 }} >> $GITHUB_ENV
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:69]: Why is it put to ENV vars if is at the output?

File: .github/workflows/build_push_service.yml [hunk 3]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+      - name: Read version file
+        # The version to be used for tagging the images is read out of the version file
+        id: read_version
+        uses: juliangruber/read-file-action@v1
+        with:
+          path: ./version
+      - name: Set env AVCHECK_VERSION
+        run: |
+          echo "Version of antivirus_service is ${{ steps.read_version.outputs.content }}"
+          echo "AVCHECK_VERSION=${{ steps.read_version.outputs.content }}" >> $GITHUB_ENV
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:58]: Why is it put to ENV vars if is at the output?

File: .github/workflows/build_push_service.yml [hunk 4]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+      - name: Set BRANCH_NAME
+        run: echo Set BRANCH_NAME
+        env:
+          BRANCH_NAME: ${{ steps.branch.outputs.BRANCH_NAME }}
+      - name: Read version file
+        # The version to be used for tagging the images is read out of the version file
+        id: read_version
+        uses: juliangruber/read-file-action@v1
+        with:
+          path: ./version
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:54]: Why is the version file read?

File: .github/workflows/build_push_service.yml [hunk 5]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+      - name: Assemble Dockertag for hotfix
+        if: env.BRANCH_PREFIX == 'hotfix'
+        run: |
+          echo "DOCKER_TAG_POSTFIX=hotfix_${{env.JIRA_TicketID}}_latest" >> $GITHUB_ENV
+      # master -> master_v<AVCHECK_VERSION>_latest, master_v<AVCHECK_VERSION>_GITSHA
+      - name: Assemble Dockertag for master
+        if: env.BRANCH_PREFIX == 'master'
+        run: |
+          echo "DOCKER_TAG_POSTFIX=master_${{env.AVCHECK_VERSION}}_latest" >> $GITHUB_ENV
+          echo "DOCKER_TAG_SHA=master_${{env.AVCHECK_VERSION}}_${{env.GITHUB_SHA}}" >> $GITHUB_ENV
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:107]: Too much special cases, why was they not designed as separate workflows?

File: .github/workflows/build_push_service.yml [hunk 6]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_TOKEN }}
+      - name: Build and push Scanfile
+        uses: docker/build-push-action@v2.2.1
+        with:
+          context: .
+          file: ./Dockerfile.scanfile
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANFILE_REPO }}:${{ env.DOCKER_TAG_POSTFIX }}
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:136]: why not more geneal why, with more tags then one pushed?

File: .github/workflows/build_push_service.yml [hunk 7]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_SHA }}
+      - name: Build and push SHA Tag for Master and Release on Webserver
+        if: ${{ github.event.inputs.branch-prefix == 'master' || github.event.inputs.branch-prefix == 'release' }}
+        uses: docker/build-push-action@v2.2.1
+        with:
+          context: .
+          file: ./Dockerfile.webserver
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_SHA }}
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:174]: To mutch steps with to offen build, why where the SHA tags not push with this bevore?

File: .github/workflows/build_push_service.yml [hunk 8]
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
...
 on:
+  push:
+    branches:
+      - develop
+      - master
+      # release branches are of e.g. release/25.3.0
+      - release/[0-9]+.[0-9]+.[0-9]+
+      # OPS-1664 naming convention <branch>/<JIRA-Ticket ID>-<Jira_Summary>
+      - hotfix/[A-Z]+-[0-9]+-[a-zA-Z0-9]+_[a-zA-Z0-9]+*
+      - feature/[A-Z]+-[0-9]+-[a-zA-Z0-9]+_[a-zA-Z0-9]+*
```
Comment (mamutmk5) [inline .github/workflows/build_push_service.yml:17]: Why are it on so many cases at once, it's created many special cases to handle.

File: version [hunk 9]
```diff
@@ -0,0 +1 @@
+1.0.0
```
Comment (mamutmk5) [inline version:1]: Why is a version file add what not used by python and why weren't the git tag used for versioning?

**Review by mamutmk5 — CHANGES_REQUESTED** [review]
On my opinion is to complex designed with too many special cases.

File: .github/workflows/build_push_service.yml (hunk 1 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:81]: So other actions have access to env vars, we need this var to build correct dockertag with new naming convention.

File: .github/workflows/build_push_service.yml (hunk 2 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:69]: Other actions have access to env vars, we need this var to build correct dockertag with new naming convention.

File: .github/workflows/build_push_service.yml (hunk 3 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:58]: Other actions have access to env vars, we need this var to build correct dockertag with new naming convention.
see EPIC OPS-1499 Provide team based HPI Schul-Cloud instances
new naming convention confluence: Naming Convention on Builds https://docs.hpi-schul-cloud.org/display/TSC/Naming+Convention+on+Builds

File: .github/workflows/build_push_service.yml (hunk 4 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:54]: need for new naming convention dockertag master/ release

File: .github/workflows/build_push_service.yml (hunk 5 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:107]: See epic, to meet naming convention it is necessary.

Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:107]: See epic, to meet naming convention it is necessary. We will build all branches with one Workflow.
One thing was that actions have very little capability's on conditions. It doesn't look smart but now it works.

File: .github/workflows/build_push_service.yml (hunk 6 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:136]: I see three steps for three docker images to be build. would be grateful for easier example.

File: .github/workflows/build_push_service.yml (hunk 7 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:174]: 1. conditions in actions are limited. So 5 braches with 3 images.
2. we had SHA tags before an dev's and devops demanded to keep SHA Tags and enhanced epic / naming convention.

File: .github/workflows/build_push_service.yml (hunk 8 above)
Comment (RalfNellesCap) [reply .github/workflows/build_push_service.yml:17]: No, Workflow should just fail on old naming scheme.

File: version (hunk 9 above)
Comment (RalfNellesCap) [reply version:1]: python should be the next step with our own python action for all sc-apps
See: feature/OPS-1722-build_and_deploy_antivirus

File: .github/workflows/build_push_service.yml (hunk 4 above)
Comment (mamutmk5) [reply .github/workflows/build_push_service.yml:54]: what is with the version at setup.py and how do it use the ./version file?

https://github.com/hpi-schul-cloud/antivirus_check_service/blob/feature/OPS-1722-build_and_deploy_antivirus/setup.py

File: .github/workflows/build_push_service.yml (hunk 1 above)
Comment (mamutmk5) [reply .github/workflows/build_push_service.yml:81]: okay, then my question where not so good to understand.
Why do you put the output form an named step to the env vars again and not uesed the step output later?
The steps.branch_prefix.outputs.group1 is allrady an env var.....

**Comment by RalfNellesCap** [comment]
- Added SHA tag to all branches
- fixed routing through branches with new conditions to build dockertags
- fixed double versioning files, joint to single point of truth.

File: version (hunk 9 above)
Comment (RalfNellesCap) [reply version:1]: version file now is used by python the git tag uses the version text file for versioning.

**Comment by RalfNellesCap** [comment]
- unified version
- added SHA Tags to all branches
- cleaned up conditions-steps
- action runs
- pushes tag SHA
- pushes three images with 2 Tags each.

Asked for review and merge.

**Review by mamutmk5 — APPROVED** [review]

**Review by ldahlke — APPROVED** [review]

**Pull request merged**